src/interpreter.rs

```diff
@@ -1924,27 +1924,26 @@ impl<'source> Interpreter<'source> {
                 head: rule_head,
                 bodies: rule_body,
             } => {
-                if matches!(rule_head, RuleHead::Func { .. }) {
-                    return Ok(());
-                }
+                if !matches!(rule_head, RuleHead::Func { .. }) {
```

Comment:
Good catch. Previously, the state wasn't being managed cleanly on early exit.

We'd want to move line 1952 into this block so that only non function rules are marked as processed.